### PR TITLE
Saurabh.ajmera/rtb 954 with cte code refactoring via with statement algebra builder

### PR DESCRIPTION
**Pull request by sauajmera**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/rel2sql/CTERelToSqlUtil.java [hunk 1]
```diff
@@ -137,34 +175,137 @@ public static void addSqlWithItemNode(SqlWithItem sqlWithItem, List<SqlNode> sql
...
-        updateNode(rightNode);
-
-        //if left and right is join again need recursive call
-      } else if (sqlSelect.getFrom() instanceof SqlBasicCall) {
-        updateNode(sqlSelect.getFrom());
-      } else if (sqlSelect.getFrom() instanceof SqlWithItem) {
-        sqlSelect.setFrom(((SqlWithItem) sqlSelect.getFrom()).name);
+      if (sqlNode instanceof SqlSelect) {
+        SqlSelect sqlSelect = (SqlSelect) sqlNode;
+        // Handle FROM clause
```
Comment (rushikeshsapate-dm) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/CTERelToSqlUtil.java:180]: can we remove these single line comments as code is self explanatory

File: core/src/main/java/org/apache/calcite/plan/TableAliasTrait.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.plan;
+
+/**
+ * TableAliasTrait is used to identify if a given rel has a TableAliasTrait.
+ */
+public class TableAliasTrait implements RelTrait {
```
Comment (rushikeshsapate-dm) [inline core/src/main/java/org/apache/calcite/plan/TableAliasTrait.java:22]: add TCs around changes

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java [hunk 3]
```diff
@@ -2949,6 +2956,9 @@ private String getTableName(String alias, RelNode rel) {
     } else {
       tableName = alias;
     }
+    if (tableName == null && alias != null) {
```
Comment (Abhi-210793) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:2959]: Why is this check required if in else part we are always setting - tableName = alias;

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java [hunk 4]
```diff
@@ -2320,12 +2326,10 @@ && hasNested((Aggregate) rel, SqlImplementor::isAggregate)) {
       }
 
       if (rel instanceof Project && rel.getInput(0) instanceof Aggregate) {
-        if (CTERelToSqlUtil.isCteScopeTrait(rel.getTraitSet())
-            || CTERelToSqlUtil.isCteDefinationTrait(rel.getTraitSet())) {
+        if (CTERelToSqlUtil.isCTEScopeOrDefinitionTrait(rel.getTraitSet())) {
```
Comment (Abhi-210793) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:2329]: Can you please merge these two if conditions

File: core/src/main/java/org/apache/calcite/rel/rel2sql/CTERelToSqlUtil.java [hunk 5]
```diff
@@ -137,34 +175,137 @@ public static void addSqlWithItemNode(SqlWithItem sqlWithItem, List<SqlNode> sql
...
+    SqlNode query = withItem.query;
+
+    if (query instanceof SqlSelect) {
+      updateSqlNode(query);
+    } else if (query instanceof SqlBasicCall) {
+      processBasicCall(query);
+    }
+  }
+
+  public static void updateSqlJoinNode(SqlJoin sqlJoin) {
```
Comment (Abhi-210793) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/CTERelToSqlUtil.java:246]: This method can be simplified like this -
`
public static void updateSqlJoinNode(SqlJoin sqlJoin) {
    updateNodeOrJoin(sqlJoin.getLeft());
    updateNodeOrJoin(sqlJoin.getRight());
}

private static void updateNodeOrJoin(SqlNode node) {
    if (node instanceof SqlJoin) {
        updateSqlJoinNode((SqlJoin) node);  
    } else if (node instanceof SqlBasicCall) {
        updateNode(node); 
    } else {
        updateSqlNode(node);  
    }
}
`

File: core/src/main/java/org/apache/calcite/rel/rel2sql/CTERelToSqlUtil.java [hunk 6]
```diff
@@ -137,34 +175,137 @@ public static void addSqlWithItemNode(SqlWithItem sqlWithItem, List<SqlNode> sql
...
+        if (isNestedCte(fromNode)
+            &&
+            fromNode instanceof SqlBasicCall
+            &&
+            ((SqlBasicCall) fromNode).getOperator() instanceof SqlAsOperator) {
+          updateNode(fromNode);
+        }
+        // Handle WHERE clause
+        SqlNode whereNode = sqlSelect.getWhere();
+        if (whereNode != null && whereNode instanceof SqlBasicCall) {
```
Comment (Abhi-210793) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/CTERelToSqlUtil.java:192]: Will this check `whereNode instanceof SqlBasicCall` not be sufficient?? Do we need null check

**Review by Abhi-210793 — APPROVED** [review]
Review done

**Review by dm-shantanuc — APPROVED** [review]

**Pull request merged**